src/lib/flight_tasks/tasks/AutoLineSmoothVel/TrajectoryConstraints.hpp

```diff
@@ -126,30 +126,50 @@ float computeXYSpeedFromWaypoints(const Vector3f waypoints[N], const VehicleDyna
 	return max_speed;
 }
 
-inline bool clampToXYNorm(Vector3f &target, float max_xy_norm)
+/*
+ * Constrain the 3D vector given a maximum XY norm
+ * If the XY norm of the 3D vector is larger than the maximum norm, the whole vector
+ * is scaled down to respect the constraint.
+ * If the maximum norm is small (defined by the "accuracy" parameter),
+ * only the XY components are scaled down to avoid affecting
+ * Z in case of numerical issues
+ */
+inline void clampToXYNorm(Vector3f &target, float max_xy_norm, float accuracy = FLT_EPSILON)
 {
 	const float xynorm = target.xy().norm();
-	const float scale_factor = max_xy_norm / xynorm;
+	const float scale_factor = (xynorm > FLT_EPSILON)
+				   ? max_xy_norm / xynorm
+				   : 1.f;
 
-	if (scale_factor < 1 && PX4_ISFINITE(scale_factor) && xynorm > FLT_EPSILON) {
+	if (max_xy_norm < accuracy && xynorm < accuracy) {
+		target.xy() = Vector2f(target) * scale_factor;
+
+	} else if (scale_factor < 1.f && PX4_ISFINITE(scale_factor) && xynorm > FLT_EPSILON) {
```

Comment:
Don't need this anymore (or below), and probably move the PX4_ISFINITE check into the scale factor calculation, just need to check max_xy_norm.

```suggestion
	} else if (scale_factor < 1.f ) {
```